project4/name_server/nameserver.cpp

```diff
@@ -1,41 +1,276 @@
 #include "nameserver.h"
+
 namespace nameserver {
+
 Nameserver::Nameserver(
     std::shared_ptr<thread_pool::ThreadPool> pool,
     std::shared_ptr<nameserver::tasks::ConsoleTask> console_task,
-    const std::filesystem::path config_file) {
+    int port,
+    message_passing::Endpoint bootstrap) {
+  port_ = port;
+  bootstrap_ = bootstrap;
   console_task_ = console_task;
   threadpool_ = pool;
+  client_ = std::make_unique<message_passing::Client>(threadpool_, bootstrap_);
+  server_ = std::make_unique<message_passing::Server>(threadpool_, port_);
+}
+
+void Nameserver::HandleRequest(consistent_hash_msgs::NameServerMessage& msg,
+                               message_passing::Endpoint source) {
+  if (msg.has_update_succ_req()) {
+    auto req = msg.update_succ_req();
+    HandleRequest(req);
+  } else if (msg.has_delete_result()) {
+    auto req = msg.delete_result();
+    HandleRequest(req);
+  } else if (msg.has_look_up_result()) {
+    auto req = msg.look_up_result();
+    HandleRequest(req);
+  } else if (msg.has_insert_result()) {
+    auto req = msg.insert_result();
+    HandleRequest(req);
+  } else if (msg.has_entrance_info()) {
+    auto req = msg.entrance_info();
+    HandleRequest(req);
+  } else if (msg.has_exit_info()) {
+    HandleRequest(msg.exit_info());
+  } else if (msg.has_update_pred_req()) {
+    auto req = msg.update_pred_req();
+    HandleRequest(req, source);
+  }
 }
 
-bool Nameserver::Enter() { return false; }
+bool Nameserver::Enter() {
+  consistent_hash_msgs::EntranceRequest entrance_req;
+  consistent_hash_msgs::EntranceInformation entrance_info;
+  consistent_hash_msgs::UpdatePredecessorResponse update_pred_res;
+  consistent_hash_msgs::UpdatePredecessorRequest update_pred_req;
+  consistent_hash_msgs::UpdateSuccessorRequest update_succ_req;
+
+  // send entrance info request to bootstrap
+  if (!client_->SendRequest(entrance_req, &entrance_info)){ return false; }
```

Comment:
This is not safe, because `client_` seems to get reset every time you want to connect to a new one. I was going to comment about this somewhere else, but in general, I think there's not much point in making this a class member, since the predecessor and successor can change. Just use local `Client` instances when you want to send something. Then you can ditch the `unique_ptr`s too. The only place where it might be reasonable to have a `Client` as a class member is in the case of the `Client` connected to the bootstrap node, since that's fixed.